### PR TITLE
feat: working openrouter support

### PR DESCRIPTION
**Pull request by cpacker**
Testing OpenRouter support via overriding OPENAI_API_KEY and OPENAPI_BASE_URL.

Seems to be working:
```sh
% letta run                         

? Would you like to select an existing agent? No

🧬 Creating new agent...
? Select LLM model: (Use arrow keys)
 » memgpt-openai
   meta-llama/llama-3.2-11b-vision-instruct:free
   meta-llama/llama-3.2-1b-instruct:free
   meta-llama/llama-3.2-3b-instruct:free
   meta-llama/llama-3.1-8b-instruct:free
   google/gemini-pro-1.5-exp
   google/gemini-flash-1.5-exp
   google/gemini-flash-1.5-8b-exp
   meta-llama/llama-3.2-1b-instruct
   meta-llama/llama-3.2-3b-instruct
   google/gemini-flash-1.5-8b
   mistralai/mistral-7b-instruct
   mistralai/mistral-7b-instruct-v0.3
   meta-llama/llama-3-8b-instruct
   meta-llama/llama-3.1-8b-instruct
   meta-llama/llama-3.2-11b-vision-instruct
   google/gemini-flash-1.5
...
```

Definitely suffers from the empty content field problem, so we need @mattzh72 's refactor to enable setting `inner_thoughts_in_kwargs` more easily:
```
% letta run                         

? Would you like to select an existing agent? No

🧬 Creating new agent...
? Select LLM model: mistralai/mixtral-8x7b-instruct
? Select embedding model: BAAI/bge-large-en-v1.5
->  🤖 Using persona profile: 'sam_pov'
->  🧑 Using human profile: 'basic'
->  🛠️  8 tools: send_message, pause_heartbeats, conversation_search, conversation_search_date, archival_memory_insert, archival_memory_search, core_memory_append, core_memory_replace
🎉 Created new agent 'BountifulYak' (id=agent-25dc8d2c-1e8c-4631-9cf1-7ef49c98f0a8)

Hit enter to begin (will request first Letta message)


🤖 Hello, Chad. It's nice to meet you. Now, let's chat. 
What keeps you busy these days? Any hobbies or passions you'd like to share?

> Enter your message:
```

Fine for OAI:
```
% letta run

? Would you like to select an existing agent? No

🧬 Creating new agent...
? Select LLM model: openai/gpt-4o-mini-2024-07-18
? Select embedding model: BAAI/bge-large-en-v1.5
->  🤖 Using persona profile: 'sam_pov'
->  🧑 Using human profile: 'basic'
->  🛠️  8 tools: send_message, pause_heartbeats, conversation_search, conversation_search_date, archival_memory_insert, archival_memory_search, core_memory_append, core_memory_replace
🎉 Created new agent 'AmazingBiscuit' (id=agent-85c55469-06be-48f1-b6cf-57cf31c1d466)

Hit enter to begin (will request first Letta message)


💭 User has logged in for the first time. I want to make this a memorable experience.
🤖 Welcome! It's great to have you here. How are you feeling about your first login?

> Enter your message:
```

### CODE REVIEW

File: letta/providers.py [hunk 1]
```diff
@@ -53,17 +53,28 @@ def list_embedding_models(self):
 class OpenAIProvider(Provider):
     name: str = "openai"
     api_key: str = Field(..., description="API key for the OpenAI API.")
-    base_url: str = "https://api.openai.com/v1"
+    base_url: str = Field(..., description="Base URL for the OpenAI API.")
```
Comment (cpacker) [inline letta/providers.py:56]: Note: allow changes to `base_url`

File: letta/providers.py [hunk 2]
```diff
@@ -53,17 +53,28 @@ def list_embedding_models(self):
...
 
     def list_llm_models(self) -> List[LLMConfig]:
         from letta.llm_api.openai import openai_get_model_list
 
-        response = openai_get_model_list(self.base_url, api_key=self.api_key)
-        model_options = [obj["id"] for obj in response["data"]]
+        # Some hardcoded support for OpenRouter (so that we only get models with tool calling support)...
+        # See: https://openrouter.ai/docs/requests
+        extra_params = {"supported_parameters": "tools"} if "openrouter.ai" in self.base_url else None
```
Comment (cpacker) [inline letta/providers.py:63]: This is kind of special case code for OpenRouter, but I'm not sure of a clean way to do this other than to make a separate Provider class for OpenRouter? But that seems excessive?

File: letta/server/server.py [hunk 3]
```diff
@@ -261,9 +262,9 @@ def __init__(
         self.add_default_tools(module_name="base")
 
         # collect providers (always has Letta as a default)
-        self._enabled_providers = [LettaProvider()]
+        self._enabled_providers: List[Provider] = [LettaProvider()]
```
Comment (cpacker) [inline letta/server/server.py:265]: This fixes typing warnings on pylance

File: letta/settings.py [hunk 4]
```diff
@@ -11,7 +11,7 @@ class ModelSettings(BaseSettings):
 
     # openai
     openai_api_key: Optional[str] = None
-    # TODO: provide overriding BASE_URL?
+    openai_base_url: Optional[str] = "https://api.openai.com/v1"
```
Comment (cpacker) [inline letta/settings.py:14]: The default for the base url is now pulled up here.

Comment (cpacker) [reply letta/settings.py:14]: also FYI I think this env var is usually `OPENAI_API_BASE`, not `OPENAI_API_BASE_URL`, but I'm not sure how to change this in the pydantic settings field other than renaming the var to `openai_base`?

I guess it's also fine for us to use a different env var? idk

eg https://github.com/langchain-ai/langchain/issues/10027

Comment (sarahwooders) [reply letta/settings.py:14]: oh yeah you can just change it to be `openai_api_base` here - and we can be consistent with the other providers

**Comment by cpacker** [comment]
@sarahwooders looks like this now:
```
% letta run

? Would you like to select an existing agent? No

🧬 Creating new agent...
? Select LLM model: letta-free (https://inference.memgpt.ai)
? Select embedding model: letta-free (https://embeddings.memgpt.ai)
->  🤖 Using persona profile: 'sam_pov'
->  🧑 Using human profile: 'basic'
->  🛠️  8 tools: send_message, pause_heartbeats, conversation_search, conversation_search_date, archival_memory_insert, archival_memory_search, core_memory_append, core_memory_replace
🎉 Created new agent 'GenerousZephyr' (id=agent-06f000c3-8034-4c3c-a720-1265378df43f)

Hit enter to begin (will request first Letta message)
```

File: letta/settings.py (hunk 4 above)
Comment (cpacker) [reply letta/settings.py]: done

**Comment by andychert** [comment]
Could you show an example of how to work with openrouter + letta?

**Pull request merged**